Issue 24687: Transfer of PerformanceSpec

kelebra opened the pull request:
Partially addresses issue #24687 with transfer of `PerformanceSpec`. 

Help is appreciated with transfer of `"have some reasonable throughput under failure conditions"` as it looks like `PersistentBahavior` and `PersistentActor` handle exceptions differently. `PersistentActor` actually resumes without losing messages but `PersistentBahavior` loses its messages even with supervision strategy resume.

kelebra committed:
Issue 24687: Transfer of PerformanceSpec

patriknw commented on akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PerformanceSpec.scala:
```diff
@@ -0,0 +1,138 @@
+  }
+
+  def eventSourcedTestPersistenceBehavior(name: String, probe: TestProbe[Command]) =
+    behavior(name, probe) {
```
match may not be exhaustive.
[warn] It would fail on the following inputs: (FailAt(_), _), (StopMeasure, _)

kelebra replied:
Fixed

patriknw commented on akka-persistence-typed/src/test/scala/akka/persistence/typed/scaladsl/PerformanceSpec.scala:
```diff
@@ -0,0 +1,138 @@
+    //    FIXME: PersistentActor survives exception but persistent behavior does not, losing pending messages
+    //    "have some reasonable throughput under failure conditions" in {
+    //      stressEventSourcedPersistentActor(Some(loadCycles / 10))
+    //    }
```
I think the reason is that typed actors are stopped by default when exception is thrown and untyped are restarted (keeping same mailbox). You can try to add `Behaviors.supervise` with `SupervisorStrategy.restart`.

Ah, you already tried with `resume`.

When running it with `restart` and enabling the stack trace for the `TE` I see this:
```
[ERROR] [03/26/2018 14:28:33.045] [PerformanceSpec-akka.actor.default-dispatcher-4] [akka://PerformanceSpec/user/PerformanceSpec-44f0b7fe-ef2d-4a7c-9c1c-f9ad88fcc2c8][persistenceId:PerformanceSpec-44f0b7fe-ef2d-4a7c-9c1c-f9ad88fcc2c8][phase:persist-evts] boom
akka.testkit.typed.TE: boom
	at akka.persistence.typed.scaladsl.PerformanceSpec$$anonfun$eventSourcedTestPersistenceBehavior$1$$anonfun$apply$2.apply$mcV$sp(PerformanceSpec.scala:88)
	at akka.persistence.typed.scaladsl.Effect$$anonfun$andThen$1.apply(Effect.scala:73)
	at akka.persistence.typed.scaladsl.Effect$$anonfun$andThen$1.apply(Effect.scala:73)
	at akka.persistence.typed.internal.EventsourcedRunning.applySideEffect(EventsourcedRunning.scala:305)
	at akka.persistence.typed.internal.EventsourcedRunning.applySideEffects(EventsourcedRunning.scala:292)
	at akka.persistence.typed.internal.EventsourcedRunning$PersistingEvents.onWriteResponse$1(EventsourcedRunning.scala:209)
	at akka.persistence.typed.internal.EventsourcedRunning$PersistingEvents.onJournalResponse(EventsourcedRunning.scala:216)
	at akka.persistence.typed.internal.EventsourcedRunning$PersistingEvents.onMessage(EventsourcedRunning.scala:183)
	at akka.persistence.typed.internal.EventsourcedRunning$PersistingEvents.onMessage(EventsourcedRunning.scala:171)
	at akka.actor.typed.scaladsl.MutableBehavior.receive(MutableBehavior.scala:23)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:170)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Intercept.receive(BehaviorImpl.scala:173)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.BehaviorImpl$Widened.receive(BehaviorImpl.scala:52)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.Supervisor.receive(Restarter.scala:96)
	at akka.actor.typed.Behavior$.interpret(Behavior.scala:341)
	at akka.actor.typed.Behavior$.interpretMessage(Behavior.scala:318)
	at akka.actor.typed.internal.adapter.ActorAdapter.akka$actor$typed$internal$adapter$ActorAdapter$$handleMessage(ActorAdapter.scala:64)
	at akka.actor.typed.internal.adapter.ActorAdapter$$anonfun$running$1.applyOrElse(ActorAdapter.scala:60)
	at akka.actor.Actor$class.aroundReceive(Actor.scala:517)
	at akka.actor.typed.internal.adapter.ActorAdapter.aroundReceive(ActorAdapter.scala:19)
	at akka.actor.ActorCell.receiveMessage(ActorCell.scala:588)
	at akka.actor.ActorCell.invoke(ActorCell.scala:557)
	at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:258)
	at akka.dispatch.Mailbox.run(Mailbox.scala:225)
	at akka.dispatch.Mailbox.exec(Mailbox.scala:235)
	at akka.dispatch.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
	at akka.dispatch.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
	at akka.dispatch.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
	at akka.dispatch.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

I wonder what that recursion can be?

I have to look at this more later...

patriknw replied:
ok, I created issue https://github.com/akka/akka/issues/24787 for the weird stack.

The problem you are seeing here though is that the `StashBuffer` is recreated for each restart and thereby dropping messages in the stash. I'm not sure how we should solve that?

The only place we could keep the stash is in the `PersistentBehaviorImpl` but that could potentially be used for multiple `spawn`, but in practice that might never be possible since there can only be one instance with the same `persistenceId`. On the other hand it can be spawned again after `stop` and then the stash would survive that also which is not desired. No that's not good.

patriknw replied:
Second thought, we can clean the stash in a PostStop signal if we hold on to the `StashBuffer` in `PersistentBehaviorImpl`. Perhaps that is the only way.

@kelebra Do you understand what I mean with the `StashBuffer` here and could you try to implement that idea of only creating the buffer once, and then use the same for subsequent calls to `PersistentBehaviorImpl.apply`?

kelebra replied:
@patriknw, thanks a lot for explanation. So far as I understand that we create new instance of `StashBuffer` in `EventsourcedSetup.apply` which in turn happens during `PersistentBehaviorImpl.apply`. If `PersistentBehaviorImpl.apply` is called one more time for resume and/or restart we need to keep stash in order not to loose commands of internal protocol.

So if I understand you right I should do the following:

* Move `internalStash` creation out of `EventsourcedSetup` to `PersistentBehaviorImpl`
* Create `EventsourcedSetup` in `PersistentBehaviorImpl.apply` using `internalStash` from `PersistentBehaviorImpl` (as per first step)

Will try to add changes to this PR today itself for your review.

kelebra replied:
I wonder which behaviors out of four implementations should be impacted or to specify question which sub-behaviors even should be impacted?

kelebra replied:
Introduced `akka.persistence.typed.internal.EventsourcedStashReferenceManagement` that holds reference and prevents stash from being recreated on `apply` and mixed it into `PersistentBehaviorImpl` for creation using inherited `stash`. Uncommented `"have some reasonable throughput under failure conditions"` that now passes.

kelebra committed:
Issue 24687: Added long living stack reference, uncommented test.

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedStashReferenceManagement.scala:
```diff
@@ -0,0 +1,20 @@
+  private var internal: StashBuffer[InternalProtocol] = _
+
+  def stash(settings: EventsourcedSettings): StashBuffer[InternalProtocol] = {
+    if (internal == null) internal = StashBuffer(settings.stashCapacity)
```
can use `akka.util.OptionVal` instead of `null`, it's allocation free, and match is also allocation free

kelebra replied:
Fixed

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedStashReferenceManagement.scala:
```diff
@@ -0,0 +1,20 @@
+
+  private var internal: StashBuffer[InternalProtocol] = _
+
+  def stash(settings: EventsourcedSettings): StashBuffer[InternalProtocol] = {
```
perhaps rename to `stashBuffer`

kelebra replied:
Fixed

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/PersistentBehaviorImpl.scala:
```diff
@@ -45,7 +45,7 @@ private[akka] final case class PersistentBehaviorImpl[Command, Event, State](
           holdingRecoveryPermit = false,
           settings = settings,
-          internalStash = StashBuffer(settings.stashCapacity)
+          internalStash = stash(settings)
```
ok, let's see how we can add the `PostStop` cleanup.
I don't think we can use `Behaviors.receiveSignal`, because that will consume the signal and not propagate to inner behaviors that might also need it.

Perhaps via `Behaviors.intercept`.

```
Behaviors.setup[EventsourcedBehavior.InternalProtocol] { ctx ⇒
  Behaviors.withTimers { timers ⇒
    Behaviors.intercept(...
```

kelebra replied:
I guess you meant `Behaviors.tap`? Yeah, makes sense. Will add changes to this PR.

kelebra replied:
Done

patriknw reviewed (approved):
LGTM, thanks

patriknw commented:
I fixed the merge conflict

kelebra commented:
Can we get one more `LGTM` on this one? :)

patriknw commented:
@kelebra can I ask you to rebase and squash this manually to get rid of my merge commit. "Squash and merge" in UI is scary stupid, using wrong author sometimes.

kelebra commented:
Will do, no problems

kelebra committed:
Merge branch 'master' into 24687/performance-spec

kelebra commented:
Seems like done and history rewritten. Please let me know if you want me to squash these 3 remaining commits.

johanandren commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/PersistentBehaviorImpl.scala:
```diff
@@ -25,32 +25,39 @@ private[akka] final case class PersistentBehaviorImpl[Command, Event, State](
-          settings = settings,
-          internalStash = StashBuffer(settings.stashCapacity)
-        )
+        {
```
What is this extra block?

kelebra replied:
Online merging is evil :sweat:. Noted for the future and removed redundant block.

johanandren commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/EventsourcedStashReferenceManagement.scala:
```diff
@@ -0,0 +1,35 @@
+      }
+    case _ ⇒ Unit
+  }
+}
```
I don't think this is right, It doesn't align with the general design of the new typed actor as it mixes in mutable state (like how it was done in the old persistent actor) into an object that is actually replaced with another behavior pretty quickly. I think the state should be completely moved into setup.

patriknw replied:
The problem is that the stash must survive a restart. Thoughts about design here https://github.com/akka/akka/pull/24782#comment-177073511

I agree though that it is questionable. Let’s wait with merging this, perhaps someone has some better brilliant idea.

johanandren replied:
Ah, I see, I missed that comment before. If we are to keep this solution I think there should be a comment explaining the why in the sources.

patriknw replied:
If you are in control of the whole chain, Behaviors.setup -> Behaviors.supervise -> Behaviors.immutable, it's not a problem because then you can create and hold the StashBuffer to the left of the supervise. Here the user may add a supervise before our stuff.

kelebra replied:
> Let’s wait with merging this, perhaps someone has some better brilliant idea.

One thing I can think of is to manage `StashBuffer` in separate actor/behavior (smth called like `StashReferenceBehavior`) and expand `InternalProtocol` with command like `ObtainStash`/`ClearStash`/`StashObtained(stash)` to make it all more transparent. However it seems that potential behavior reference (`StashReferenceBehavior` here) would have to be managed separately which brings us to basically the same solution which is implemented now.

Fork and branch are going nowhere and in case anybody comes up with something better we could redesign solution in this PR but as of now it at least fixes pending issue with restart of persistent actor.

patriknw replied:
@johanandren any more thoughts about this?

kelebra committed:
Issue 24687: Addressed review comments

kelebra commented:
Addressed review comments:

- Added comment explaining intention behind the trait and mutable state
- Removed block introduced during cowboish online merge

patriknw commented:
PLS BUILD
